crates/extrinsics/src/contract_info.rs

```diff
@@ -32,45 +32,135 @@ use scale::Decode;
 use std::option::Option;
 use subxt::{
     backend::legacy::LegacyRpcMethods,
+    dynamic::DecodedValueThunk,
+    ext::{
+        scale_decode::DecodeAsType,
+        scale_value::Value,
+    },
+    storage::dynamic,
     utils::AccountId32,
 };
 
+/// Return the account data for an account ID.
+async fn get_account_balance(
+    account: &AccountId32,
+    rpc: &LegacyRpcMethods<DefaultConfig>,
+    client: &Client,
+) -> Result<AccountData> {
+    let storage_query =
+        subxt::dynamic::storage("System", "Account", vec![Value::from_bytes(account)]);
+    let best_block = get_best_block(rpc).await?;
+
+    let account = client
+        .storage()
+        .at(best_block)
+        .fetch(&storage_query)
+        .await?
+        .ok_or_else(|| anyhow::anyhow!("Failed to fetch account data"))?;
+
+    let data = account.as_type::<AccountInfo>()?.data;
+    Ok(data)
+}
+
 /// Fetch the contract info from the storage using the provided client.
 pub async fn fetch_contract_info(
     contract: &AccountId32,
     rpc: &LegacyRpcMethods<DefaultConfig>,
     client: &Client,
-) -> Result<Option<ContractInfo>> {
-    let info_contract_call = api::storage().contracts().contract_info_of(contract);
-
+) -> Result<ContractInfo> {
     let best_block = get_best_block(rpc).await?;
 
-    let contract_info_of = client
+    let contract_info_address = dynamic(
+        "Contracts",
+        "ContractInfoOf",
+        vec![Value::from_bytes(contract)],
+    );
+    let contract_info_value = client
         .storage()
         .at(best_block)
-        .fetch(&info_contract_call)
-        .await?;
+        .fetch(&contract_info_address)
+        .await?
+        .ok_or_else(|| {
+            anyhow!(
+                "No contract information was found for account id {}",
+                contract
+            )
+        })?;
+
+    let contract_info_raw = ContractInfoRaw::new(contract, contract_info_value)?;
+    let deposit_account = match contract_info_raw {
+        ContractInfoRaw::MainAccountDeposit(ref account, _)
+        | ContractInfoRaw::SecondaryAccountDeposit(ref account, _) => account,
+    };
+
+    let deposit_account_data = get_account_balance(deposit_account, rpc, client).await?;
+    Ok(contract_info_raw.into_contract_info(deposit_account_data))
+}
 
-    match contract_info_of {
-        Some(info_result) => {
-            let convert_trie_id = hex::encode(info_result.trie_id.0);
-            Ok(Some(ContractInfo {
-                trie_id: convert_trie_id,
-                code_hash: info_result.code_hash,
-                storage_items: info_result.storage_items,
-                storage_item_deposit: info_result.storage_item_deposit,
-            }))
+/// Enum representing different types of contract info, distinguishing between
+/// main account deposit and secondary account deposit.
+pub enum ContractInfoRaw {
```

Comment:
```suggestion
enum ContractInfoRaw {
```
Only used internally here.